### ISSUE TITLE
 Grover::JavaScript::ProtocolError: Network.enable timed out. Increase the 'protocolTimeout' setting in launch/connect calls for a higher timeout if needed.

### ISSUE THREAD

**Issue by JosephTico** [Studiosity/grover#248]
This is a very weird thing, never had any issue related to timeouts until a couple days ago and this has started happening randomly, not every request to Grover does this. Using Alpine 3.20.

My initializer:

```
# config/initializers/grover.rb
Grover.configure do |config|
  config.options = {
    format: 'letter',
    emulate_media: 'screen',
    margin: {
      top: '1cm',
      bottom: '1cm',
    },
    print_background: true,
    cache: false,
    scale: 0.75
  }
end
```

I'm using Rails 7.1, Ruby 3.3.3. For Rendering PDFs I use this helper function that renders the action to a string and passes it to Grover.

```
  def render_pdf(template, disposition: 'inline', filename: 'file')
      send_data(
        Grover.new(
          render_to_string(
            {
              action: action_name.to_s,
              formats: [:pdf],
              template:,
              layout: 'pdf'
            }
          )
        ).to_pdf, disposition:, filename: "#{filename}.pdf", type: 'application/pdf'
      )
  end
```

Any ideas? Thank you.

**Comment by CeeBeeUK** [comment]
About to ask the same question...
We have spent a few days trying different debugging issues, the only obvious thing to change was chromium incrementing from 126 to 127...

On a broken environment, opening node and running 
```

(async () => {
  const puppeteer = require("puppeteer");
  // Launch the browser and open a new blank page
  const browser = await puppeteer.launch({args:['--disable-gpu', '--no-sandbox','--disable-setuid-sandbox']});
  const page = await browser.newPage();
  const version = await page.browser().version();
  console.log("page browser version: " + version);
  await page.goto('https://bbc.co.uk/news');

  console.log("Page title: " + await page.title());

  const element = await page.waitForSelector('h1')
  let value = await element.evaluate(el => el.textContent)
  console.log('h1 = ' + value);

  await browser.close();
})();
```
returns 
```
page browser version: Chrome/127.0.6533.88
Page title: Home - BBC News
h1 = BBC News
```

But when I start a rails console and run 
```ruby
Grover.new('test').to_pdf
```
it will hang and return the error above

Running the same sequence on a working build returns the PDF output

The working infrastructure is:
Alpine 3:20
Grover 1.1.9
Puppeteer 22.13.1
Chromium 126.0.6478.182 Alpine Linux

Broken infrastructure:
Alpine 3:20
Grover 1.1.9
Puppeteer 22.13.1
Chromium 127.0.6533.72 Alpine Linux

We have updated a build this morning to use the latest `Chromium 127.0.6533.88 Alpine Linux` but it is still failing

**Comment by JosephTico** [comment]
Yeah, it looks like this is a Puppeteer issue with Alpine 3.20's Chromium version.

For now a quick fix is downgrading to Alpine 3.19 as it looks like it's using a different Chromium that is still compatible.

**Comment by CeeBeeUK** [comment]
I'm not sure it is puppeteer... if you launch puppeteer via node it successfully renders pages

But yes, rolling back to Alpine 3:19 and puppeteer 22.7.1 seems to be our only recourse to get this working

**Comment by JosephTico** [comment]
I think it could be related to this:

https://github.com/puppeteer/puppeteer/issues/11640#issuecomment-2243353372

The thing making it fail appears to be the `--single-process` flag which I assume Grover could be setting.

More issues related to this on Puppeteer's repo:

https://github.com/puppeteer/puppeteer/issues/12637
https://github.com/puppeteer/puppeteer/issues/12189

**Comment by CeeBeeUK** [comment]
I'm having some success with changing my Grover configuration launch_args
from 
```ruby
launch_args: %w[--no-sandbox]
```
to
```ruby
launch_args: %w[--disable-gpu --no-sandbox --disable-setuid-sandbox]
```
My testing is continuing but, so far, it is looking hopeful

**Comment by abrom** [comment]
Not sure I can add much to this issue, but I'd welcome an update to the README if/when there is a resolution. 

As far as I can tell from the discussion above this isn't a Grover issue, but rather a Chrome/Chromium + your environment issue. Grover does not add/inject any launch arguments that you haven't told it to.

Of course, forewarning others about potential configuration mismatches like this would be great.

**Comment by JoaoFerreiraLuz** [comment]
it work for me 
```

# BASE IMAGE CONTAINER
FROM node:18-alpine3.18 As base
RUN apk update
RUN apk add --no-cache libc6-compat
RUN apk add --no-cache bash
RUN apk add --no-cache ffmpeg
RUN apk add --no-cache python3 make g++

USER root

RUN apk update && \
    apk add --no-cache \
    libc6-compat \
    chromium \
    nss \
    freetype \
    harfbuzz \
    ca-certificates \
    ttf-freefont && \
    rm -rf /var/cache/apk/*

ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD true
ENV PUPPETEER_CACHE_DIR=/tmp/.cache
ENV PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser

RUN addgroup -S pptruser && adduser -S -G pptruser pptruser \
    && mkdir -p /home/pptruser/Downloads /app \
    && chown -R pptruser:pptruser /home/pptruser \
    && chown -R pptruser:pptruser /app

ENV PYTHON=/usr/bin/python3

```

**Comment by cameronbourgeois** [comment]
This has reared it's ugly head again for me - or at least it is a similar issue.
I have not intentionally updated any packages, it appears to be Alpine Linux have changed something so that my more recently built docker containers contain the issue.

In my production environment, about 50% of PDFs are failing. It seems to be completely random.

Alpine 3.21 (docker image ruby:3.4.3-alpine3.21)
Grover 1.2.3
Puppeteer 24.8.1
Chromium 135.0.7049.95

Using `--disable-gpu` which has been there since I last encountered this issue in Alpine 3.20.

Interestingly this issue does not seem to happen for me on my local environment, ONLY when being run either on CircleCI or AWS.
